### ISSUE TITLE
Offsets handling when restarting the control

### ISSUE THREAD

**Issue by paLeziart** [open-dynamic-robot-initiative/odri_control_interface#1]
Hello,

I am currently switching to this repository for the control of Solo12 and I encountered a weird behavior with the angular offsets of the actuators.

The first time I turn on the robot, the calibration runs properly and then the robot goes into waiting configuration until the user presses the Enter key. The function I am using retrieves the current joint position in `pos_init` then interpolates a trajectory to the waiting configuration `q_init`.

```
pos_init[:] = device.joints.positions
i = threading.Thread(target=get_input)
k = 0
i.start()
while i.is_alive():
    device.parse_sensor_data()
    if t <= T:
        x = (t / T)
        pos = x * q_init.ravel() + (1 - x) * pos_init
        t += dt
    torques = Kp_pos * (pos - device.joints.positions) - Kd_pos * device.joints.velocities
    print(k)
    print("Target:     ", pos)
    print("Joint pos: ", device.joints.positions)
    k = k+1
    device.joints.set_torques(0.0 * torques)
    device.send_command_and_wait_end_of_cycle()
```

When I turn off the control (the robot is still turned on) and launch the script again, the calibration step is not done (as expected because the robot has not been turned off) but the joints position are not correct. I get the following output, with the actuators around their "0" position.

```
0
Target:     [-0.193 -0.332 -0.036 -0.231  0.273 -0.292  0.167 -0.186  0.217  0.226  0.16  -0.347]
Joint pos:  [-0.193 -0.332 -0.036 -0.231  0.273 -0.292  0.167 -0.186  0.217  0.226  0.16  -0.347]
1
Target:     [-0.193 -0.331 -0.037 -0.231  0.273 -0.292  0.167 -0.187  0.217  0.226  0.16  -0.346]
Joint pos:  [-0.193 -0.332 -0.036 -0.231  0.273 -0.292  0.167 -0.186  0.217  0.226  0.16  -0.347]
2
Target:     [-0.193 -0.33  -0.038 -0.231  0.273 -0.293  0.167 -0.187  0.218  0.226  0.159 -0.345]
Joint pos:  [-0.193 -0.332 -0.036 -0.231  0.273 -0.292  0.167 -0.186  0.216  0.226  0.16  -0.347]
3
Target:     [-0.192 -0.33  -0.039 -0.231  0.274 -0.294  0.167 -0.187  0.219  0.226  0.158 -0.344]
Joint pos:  [-6.639e-02 -9.214e-03 -1.456e-03 -1.365e-04  9.656e-03 -1.202e-02 -1.113e-02 -1.968e-02  8.887e-03  7.803e-02  1.601e-01 -3.472e-01]
4
Target:     [-0.192 -0.329 -0.04  -0.231  0.274 -0.295  0.167 -0.188  0.22   0.225  0.158 -0.343]
Joint pos:  [-0.066 -0.009 -0.001 -0.     0.01  -0.012 -0.011 -0.02   0.009  0.078  0.009 -0.007]
5
Target:     [-0.192 -0.328 -0.041 -0.231  0.274 -0.295  0.167 -0.188  0.22   0.225  0.157 -0.341]
Joint pos:  [-0.066 -0.009 -0.001 -0.     0.01  -0.012 -0.011 -0.02   0.009  0.078  0.009 -0.007]
```

During the 3 first iterations `device.joints.positions` outputs the angular position without taking into account the offsets. The readings get correct afterward.

Am I supposed to add a waiting time when the calibration step is not needed to let the interface some time to initialize properly before starting the control?

Best,
Pierre-Alexandre

**Comment by jviereck** [comment]
Hi @paLeziart , sorry, I missed your message here. 

Is there a chance you can upload the full program you are using?

**Comment by paLeziart** [comment]
Hello @jviereck 

You can reproduce the issue with the following program : https://cloud.laas.fr/index.php/s/Fhw1rJmklobOAG7 

`import libodri_control_interface_pywrap as oci` is the compiled python .so file obtained after compiling this repository. 

With this code joint positions are correct starting from the 6th iteration:
```
PUT ON THE FLOOR.
Put the robot on the floor and press Enter
0
Target:     [-0.193 -0.332 -0.007 -0.288  0.322 -0.291  0.177 -0.177  0.217  0.217  0.161 -0.349]
Joint pos:  [-0.193 -0.332 -0.007 -0.288  0.322 -0.291  0.177 -0.177  0.217  0.217  0.161 -0.349]
1
Target:     [-0.193 -0.332 -0.008 -0.288  0.322 -0.292  0.177 -0.177  0.217  0.217  0.16  -0.348]
Joint pos:  [-0.193 -0.332 -0.007 -0.288  0.322 -0.291  0.177 -0.177  0.217  0.217  0.161 -0.349]
2
Target:     [-0.193 -0.331 -0.009 -0.288  0.322 -0.293  0.177 -0.177  0.218  0.217  0.16  -0.346]
Joint pos:  [-0.193 -0.332 -0.007 -0.288  0.322 -0.291  0.177 -0.177  0.217  0.217  0.161 -0.349]
3
Target:     [-0.192 -0.33  -0.01  -0.288  0.322 -0.293  0.177 -0.178  0.219  0.216  0.159 -0.345]
Joint pos:  [-0.193 -0.332 -0.007 -0.288  0.322 -0.291  0.177 -0.177  0.217  0.217  0.161 -0.349]
4
Target:     [-0.192 -0.329 -0.011 -0.288  0.323 -0.294  0.177 -0.178  0.22   0.216  0.158 -0.344]
Joint pos:  [-0.193 -0.009 -0.008 -0.288  0.322 -0.291  0.177 -0.177  0.217  0.217  0.161 -0.349]
5
Target:     [-0.192 -0.329 -0.012 -0.287  0.323 -0.295  0.177 -0.178  0.221  0.216  0.158 -0.343]
Joint pos:  [-6.700e-02 -9.303e-03 -8.100e-03  1.425e-04  9.225e-04 -1.165e-02 -9.297e-04 -1.054e-02  9.743e-03
  6.862e-02  1.608e-01 -3.487e-01]
6
Target:     [-0.192 -0.328 -0.013 -0.287  0.323 -0.296  0.177 -0.179  0.221  0.215  0.157 -0.342]
Joint pos:  [-0.067 -0.009 -0.008  0.     0.001 -0.012 -0.001 -0.011  0.01   0.069  0.009 -0.009]
7
Target:     [-0.191 -0.327 -0.014 -0.287  0.323 -0.296  0.177 -0.179  0.222  0.215  0.157 -0.341]
Joint pos:  [-0.067 -0.009 -0.008  0.     0.001 -0.012 -0.001 -0.011  0.01   0.069  0.009 -0.009]
```

Note that at iterations 4 and 5 some of the positions are correct (-0.009 -0.008 for 2nd and 3rd motors at iteration 4) but others are not (-0.193 for 1st motor at iteration 4).

Thank you for your help!

**Comment by jviereck** [comment]
Hi,

thanks for uploading your code.

No, there should be no delay / wait time required after the calibration finishes. Basically, the calibration applies the offset compensation in the [first step][1]. The call to `joints_->SetPositionOffsets(position_offsets_);` forces an update of the joint positions by forcing a reparsing of all the data (see [here][2]).

Therefore, I don't know why it takes a few iteration for you to see the result of the joint calibration.

Best,
Julian

[1]: https://github.com/open-dynamic-robot-initiative/odri_control_interface/blob/main/src/calibration.cpp#L75
[2]: https://github.com/open-dynamic-robot-initiative/odri_control_interface/blob/main/src/joint_modules.cpp#L200

**Comment by jviereck** [comment]
Hi @paLeziart ,

are you running your setup via a wifi connection or via a wired connection? If it's via wifi there might be some delay that could explain your measurements.

Best,
Julian

**Comment by paLeziart** [comment]
Hi @jviereck 

I am running the setup with a wired connection.

Best,
Pierre-Alexandre

**Comment by jviereck** [comment]
Hi @paLeziart ,

did you by any chance look more into this? Is this blocking you from using the library? If it is I am willing to take a closer look at the code you provided and try to reproduce this on our robots here at NYU in New York.

Best,
Julian

**Comment by paLeziart** [comment]
Hi @jviereck ,

No I did not try again since March. I will update the repository to get the recent commits then try again to see if the problem is still there.

Best,
Pierre-Alexandre

**Comment by paLeziart** [comment]
Hi @jviereck ,

During the investigation of this issue I added some print to understand what was going on and I noticed that sometimes positions returned by the interface where either Not-A-Number (with python bidings) or infinite values (with c++ code) like.

`Joints: [6.906e-310, 6.906e-310,          0,          0, 1.059e-153,  2.44e-154, 6.013e-154,  8.037e-95, 2.316e+251, 1.171e+214, 9.922e+247, 5.037e+180]`

I think it may be linked to the following line in the `WaitUntilReady` function.

https://github.com/open-dynamic-robot-initiative/odri_control_interface/blob/9e5efc8ea1ba0626fe7687a06b106f8d71e2a034/src/robot.cpp#L184

If motors are not ready, `!IsReady()` is true so it runs `HasError()` to check that motor positions and driver states. This returns an error since they are not ready and hence have not returned a valid position yet (I guess default values when the position vector is declared are infinite or 0). So `!HasError()` is False and it stops the `WaitUntilReady()` loop and the programme continues with wrong position values in the position vector. 

Using only `while (!IsReady())` could be enough since there is a `HasError()` check before sending a command to the robot with `SendCommand()` so I don't think we are loosing any safety by doing that. The code may be stuck in an infinite loop though if motors never get ready.

Another point I am going to check is the polarities of motors. In the .yaml you provide for solo12 you are using 

```
reverse_polarities: [
            true, false, true, true, false, false,
            true, false, true, true, false, false
        ]
```
when I had to use
```
reverse_polarities: [
            true, true, true, false, false, false,
            true, true, true, false, false, false
        ]
```
to get the right default positions of the legs. But it likely comes from me since with solopython we were using
```
self.motorSign = np.array([-1, +1, -1, -1, +1, +1,
                           -1, +1, -1, -1, +1, +1])
```
I will keep you informed both about that polarity thing and about the effect of the modification of the calibration procedure (waiting loop).

Best,
Pierre-Alexandre

**Comment by jviereck** [comment]
hi @paLeziart,

thanks for looking into this.

> This returns an error since they are not ready and hence have not returned a valid position yet (I guess default values when the position vector is declared are infinite or 0). So !HasError()

From reading the `HasError` code, you should see an `ERROR: Above joint limits` or `ERROR: Below joint limits` error when you have the infinite numbers for the joint position array at the beginning. Do you see this?

At the moment we don't initialize the values for the position and velocities. This might be the problem and would explain why you are getting infinite numbers.

The fix would be to call

```c++
positions_.fill(0.);
velocities_.fill(0.);
```

here:

    https://github.com/open-dynamic-robot-initiative/odri_control_interface/blob/main/src/joint_modules.cpp#L59

Can you maybe add these lines and see if you are still getting infinite values?

If possible I would like to keep the `!IsReady() && !HasError()` on the while loop. Terminating the while-loop on an error is important I think.

Let me know how it goes and if you want I am happy to dig into things as well.

Best,
Julian

**Comment by paLeziart** [comment]
Hi @jviereck ,

Thank you for your answer, I will try to fill with zeros and see if I still have infinite values.

I looked a bit into the polarity matter.

In config_solo12.yaml are you using 
```
reverse_polarities: [
            true, false, true, true, false, false,
            true, false, true, true, false, false
        ]
```
for your tests with solo like it is on main branch or are you using
```
reverse_polarities: [
            true, true, true, false, false, false,
            true, true, true, false, false, false
        ]
```
with this modification unpushed?

Here's my reasoning:
```
true, false, true, true, false, false,
true, false, true, true, false, false
```
seems like a direct translation of the polarities
```
-1, +1, -1, -1, +1, +1,
-1, +1, -1, -1, +1, +1
```
that we were using with solopython in solo12.py. 
However, the convention has changed between solopython and odri_control_interface.

* In solopython we have
```
def SetDesiredJointPosition(self,pos):
        for i in range(self.nb_motors):
            motor_pos = pos[self.motorToUrdf[i]] * self.gearRatioSigned[i]
            self.hardware.GetMotor(i).position_ref = motor_pos
        return
```
So motor number i receives the i-th polarity of the yaml configuration file.
`self.gearRatioSigned = self.motorSign * self.gearRatio`

* In odri_control_interface we have
```
void JointModules::SetDesiredPositions(ConstRefVectorXd desired_positions)
{
    for (int i = 0; i < n_; i++)
    {
        motors_[i]->SetPositionReference(polarities_[i] * desired_positions[i] * gear_ratios_[i]);
    }
}
```
With 
```
for (int i = 0; i < n_; i++)
{
    motors_.push_back(robot_if_->GetMotor(motor_numbers[i]));
    polarities_(i) = reverse_polarities(i) ? -1 : 1;
}
```
So motor number `motor_numbers[i]` receives the i-th polarity of the yaml configuration file.
`const YAML::Node& motor_numbers = joint_modules_yaml["motor_numbers"];`

* If I have a look at the effect of this change of convention on the polarities, starting from solopython.
```
-1, +1, -1, -1, +1, +1,
-1, +1, -1, -1, +1, +1
```
Once reordered with `motor_numbers: [0, 3, 2, 1, 5, 4, 6, 9, 8, 7, 11, 10]` it becomes
```
-1, -1, -1, +1, +1, +1
-1, -1, -1, +1, +1, +1
```
or in boolean
```
reverse_polarities: [
            true, true, true, false, false, false,
            true, true, true, false, false, false
        ]
```
which are the values I had to use on our solo to get the rotations in the correct directions.

If you are really using
```
true, false, true, true, false, false,
true, false, true, true, false, false
```
then there is something I don't understand because your polarities have not changed despite the change of convention in the code, so you should have some motors rotating in the wrong direction. Have I missed something that explains why the polarities have not changed on main branch?

Best,
Pierre-Alexandre

**Comment by jviereck** [comment]
Hi Pierre-Alexandre,

right now I am using the polarities for solo12 as defined in this file:

  https://github.com/open-dynamic-robot-initiative/solo/blob/master/src/solo12.cpp#L91

So I am not the `config_solo12.yaml` right now. In the `solo12.cpp` file we are using the polarities

```
  false, true, true, true, false, false, 
  false, true, true, true, false, false
```

while in the yaml file we use the polarities

```
  true, false, true, true, false, false,
  true, false, true, true, false, false
```

while the motor numbers are the same.

As we use the `solo12.cpp` in production on the real robot and I am 100% sure the angles are correct, I guess the `config_solo12.yaml` is wrong.

I will be in the lab soon and double check things on the real robot, but I am sort of sure the `config_solo12.yaml` file should change.

Sorry about that.

Best,
Julian

**Comment by paLeziart** [comment]
Issues have been fixed by #8 and #11, closing the issue.
